### ISSUE TITLE
Inaccurate Retrieval of Most Recent Glucose Data from Freestyle Libre Sensors

### ISSUE THREAD

**Issue by ArunM03** [galekseev/libre-sensor#4]
I am developing an Android application that utilizes your library to retrieve glucose history data from Freestyle Libre sensors. While the library successfully retrieves historical data, it fails to obtain the most recent glucose readings. This issue compromises the effectiveness of the app in providing timely glucose monitoring for users.

Please investigate this issue and provide guidance or an update to the library to ensure that the most recent glucose readings are accurately retrieved. This enhancement is crucial for the functionality and reliability of the application. 

![Screenshot (214)](https://github.com/galekseev/libre-sensor/assets/63395110/fa44e89b-ba37-4065-b24f-eb48a4051481)

so you can see the history variable, I'm getting the last element from the array List, but its not the most recent one, I have created a graph by using all the history values, its almost same with the official app except the latest data , so I think there is a small issue here, I hope you can fix it, Thank you

**Comment by alishah2014** [comment]
What is the frequency at which the sensor captures glucose value? Every 15 minute?

**Comment by galekseev** [comment]
Hi, @alishah2014.
The values are captured every minute. But the storage stores
- 16 trend data blocks for
-- the current value and
-- fifteen values for the last fifteen minutes and
- 32 history data blocks for the last 8 hours, all fifteen minutes apart.

Here is a little bit more about it.
https://github.com/UPetersen/LibreMonitor/wiki

**Comment by galekseev** [comment]
Hi, @ArunM03 

Could you please provide 
1. a base64 encoded reading sample from which you get the wrong data
2. what output you expect
3. what output you get

> I have created a graph by using all the history values, its almost same with the official app except the latest data

The graphs can also be useful

**Comment by ArunM03** [comment]
Hello, 
Thank  you for your reply,
Here is the base64

"tagId":"bbeb580d00a007e0","data":"UuHQGAMAAAAAAAAAAAAAAAAAAAAAAAAALrwAD90DyJRZAOQDyIxZAOMDyIRZAOkDyIxZAOwDyIRZAPEDyFhZAPgDyFSZAPwDyEBZAPkDyDhZAAEEyJBZAAkEyKBZAA8EyHhZABIEyCiZABEEyCiZABoEyGhZACEEyEhZAKIDyAhcAJADyCBaAMMDyPCYAOMDyFSYAO8DyMxZAKoDyFhbAJoDyOhaALADyGRbAJoDyMBbAIoDyOhbAMQDyORaAMEDyHBaAJsDyLBaAKcDyHhZAN0DyJRZAM8EyMwbgHgEyGxbAG0EyECaACMEyBhdAPwDyHgdgC4EyDCaAE0EyNCYADEEyASZAPUDyEyYAPYDyLSXABQEyHCXANgDyHhaAJ4DyChbALYDyIgcgLoDyKCZAMMDyBCZAK8DyNSZADEbAAAbswAIJg\/YURQHloBaAO2mFIsayAQ\/aWuu+SGD8pAHAAYIAiQMQxc8"}

My app screenshot 
![WhatsApp Image 2024-07-05 at 08 09 54_d1ef627a](https://github.com/galekseev/libre-sensor/assets/63395110/d28eb0ff-ef6e-40c8-8d7e-3c25411cfb9e)

The glucose value I'm getting from the last element of history list - 98.9 mg/dL

Official app screenshot
![WhatsApp Image 2024-07-05 at 8 03 55 AM](https://github.com/galekseev/libre-sensor/assets/63395110/df84e896-9bf8-4c5b-831a-835bd858f170)

The glucose value from the official app is - 90 mg/dL

if you want I can send my app source code to you as well 

Thank you.

**Comment by galekseev** [comment]
Hi, @ArunM03.

I've got the source of confusion. 
1. The fact is that the official app doesn't use the last history point to display the last measurement. The last history point is the point that was from 1 to 15 min ago. Actually, instead, official app uses some type of algorithm to calculate the value based on the trending data array. Once I tried different types of approximation and regression algorithms to get the same result but didn't manage to figure it out. 
2. Official app uses sensor temperature information to adjust the glucose value. Again I haven't managed to find information on how it is calculated.

So, summing up, the purpose of the library to wrap low-level binary calculations to provide the pure sensor data to make data extraction easier. Each developer has freedom on how to process that data.

My advice to get values closer to the official app is to try simple regression. It wouldn't be the same but it will be pretty close in 80-90% percent of cases. Here is the example:
https://github.com/galekseev/OpenLibre/blob/84731db9172955066fe1cb2318327a7b774212a1/app/src/main/java/com/camomile/openlibre/model/PredictionData.java#L23
